WIP: Adding Accessibility doc for Core WinForms (#13871)
dotnet/docs: docs/framework/winforms/windows-forms-accessibility-improvements.md

---
title: "Windows Forms Accessibility Improvements"

> ms.date: "08/16/2019"

Reviewer: - Can you add a `description: ` entry too, it should be around 100 characters long if possible. It should not repeat what the first paragraph in the article says.

- Also the date can be updated 😄

Author: Thank your for the review! Added description, fixed date.

  
Starting with .NET Core 3.0, Windows Forms applications benefit from all the new accessibility features (introduced in .NET Framework 4.7.1 - 4.8) without additional configuration.

## Improved ListBox Accessibility support
On .NET Core, Windows Forms includes the following changes that affect the xref:System.Windows.Forms.ListBox control:
- Enabled UIA support for ListBox control.
- Improved ListBox accessibility support by adding the ScrollItem pattern to ListBox items and by enhancing the accessibility event raising and handling and Narrator navigation through the items (caps lock navigation is not correct and does not throw the navigation outside the control unintentionally).

## Improved CheckedListBox Accessibility support
On .NET Core, Windows Forms includes the following changes that affect the xref:System.Windows.Forms.CheckedListBox control:
- Corrected CheckedListBox Bounds provided by accessibility properties for entries. Improved overall ListBox and CheckedListBox accessibility (corrected pproperty values and event model).

## Improved ComboBox Accessibility support

> On .NET Core, Windows Forms includes the following changes that affect the xref:System.Windows.Forms.ComboBox control:

Reviewer: These refs need to be fixed, they are actually XML entries:

```suggestion
On .NET Core, Windows Forms includes the following changes that affect the <xref:System.Windows.Forms.ComboBox> control:
```

You can use this format for all controls referenced in this article.

Author: Fixed this, thanks!

### Recommended Action
__The changes which could be opted in or out in .NET Framework 4.7.1 - 4.8 are now enabled by default and cannot be opted out__
  
Starting with .NET Core 3.0, Windows Forms applications benefit from all the new accessibility features (introduced in .NET Framework 4.7.1 - 4.8) without additional configuration.

## Improved ListBox Accessibility support
On .NET Core, Windows Forms includes the following changes that affect the <xref:System.Windows.Forms.ListBox control>:
- Enabled UIA support for ListBox control.
- Improved ListBox accessibility support by adding the ScrollItem pattern to ListBox items and by enhancing the accessibility event raising and handling and Narrator navigation through the items (caps lock navigation is not correct and does not throw the navigation outside the control unintentionally).

## Improved CheckedListBox Accessibility support
On .NET Core, Windows Forms includes the following changes that affect the <xref:System.Windows.Forms.CheckedListBox control>:

> - Corrected CheckedListBox Bounds provided by accessibility properties for entries. Improved overall ListBox and CheckedListBox accessibility (corrected pproperty values and event model).

Reviewer 2: ```suggestion
- Corrected CheckedListBox Bounds provided by accessibility properties for entries. 
- Improved overall ListBox and CheckedListBox accessibility (corrected pproperty values and event model).
```

- Improved DataGridView background color contrast when using Windows High Contrast themes. Changed DataGridView default back color when using HC#1, HC#2, and HC Black themes.

## Improved PropertyGrid Accessibility support
On .NET Core, Windows Forms includes the following changes that affect the <xref:System.Windows.Forms.PropertyGrid control>:
- Corrected PropertyGrid Bounds provided by accessibility properties for grid entries, improved performance of bounding rectangle calculation. For now all accessibility bounds are represented correctly taking into account the bounds of entire control, its viewport and scrolls.
- Corrected accessible names and descriptions of subcontrols to not include control type names and to avoid double announcement for control type names.

## Improved ToolStrip Accessibility support
On .NET Core, Windows Forms includes the following changes that affect the <xref:System.Windows.Forms.ToolStrip control>:
- Improved navigation through ToolStrip items. Corrected ToolStrip and Menu shift-tab navigation (back-looping the menu items when shift-tab up-arrow is pressed - navigate to the bottom menu element)

> - Improved Menu accessible navigation, corrected menu accessible control types for submenus to make submenus of type 'Menu' instead of 'MenuItem'.

Reviewer 2: We have removed `Menu` and `MenuItem` types in 3.1. 
Please verify this sentence is still applicable

Reviewer: What is there in its place?

Author: At this moment there are only ToolStrip, MenuStrip and StatusStrip controls. These controls fully cover the menu bar/tool bar functionality and there is no changes in the UI.

Author: Fixed @RussKie 's review point. Thanks!

---
title: "Windows Forms Accessibility Improvements"
ms.date: "08/16/2019"
helpviewer_keywords: 
  - "Windows Forms accessibility"
  - "accessibility"

> ms.assetid: 

Reviewer 2: `author: "M-Lipin"`

> ms.assetid: 

Author: Fixed, thanks.

---

> # Accessibility improvements in Windows Forms controls for .NET Core

Author: Fixed, thanks.

> ## Scope

Reviewer 3: Delete lines 11-17; they're based on the compatibility template, which isn't relevant here.

Author: Fixed, thanks.

Major

## Version Introduced
.NET Core 3.0

## Change Description

> The Windows Forms is continuing to improve how it works with accessibility technologies to better support Windows Forms customers. These include the following changes:

Reviewer 3: ```suggestion
Windows Forms is continuing to improve how it works with accessibility technologies to better support Windows Forms customers. These include the following changes:
```

Author: Fixed, thanks.

> - Changes in various areas of interaction with accessibility client applications including Narrator.

Reviewer 3: ```suggestion
- Changes in various areas of interaction with accessibility client applications, including Narrator.
```

Author: Fixed, thanks.

- Changes in the Accessible hierarchy (improving navigation through the UI Automation tree).
- Changes in keyboard navigation.

> - [x] The quirks which were used in .NET Framework 4.7.1 - 4.8 to opt-in/out new accessibility features have been removed, so previously quirked behavior is now enabled by default.

Reviewer 3: ```suggestion

```

> ### Recommended Action

Reviewer 3: ```suggestion

```

> __The changes which could be opted in or out in .NET Framework 4.7.1 - 4.8 are now enabled by default and cannot be opted out__

Reviewer 3: ```suggestion
> [!IMPORTANT]
>  Accessibility changes made in .NET Framework 4.7.1 through .NET Framework 4.8 are included in .NET Core 3.0 and are enabled by default. The .NET Framework supported compatibility switches that allowed applications to opt out of the new accessibility behavior. .NET Core, on the other hand, does not support these settings and does not allow applications to opt out of accessibility behavior.
```

> Applications now benefit from all the new accessibility features (introduced in .NET Framework 4.7.1 - 4.8) without additional configuration.

Reviewer 3: ```suggestion
Starting with .NET Core 3.0, Windows Forms applications benefit from all the new accessibility features (introduced in .NET Framework 4.7.1 through .NET Framework 4.8 without additional configuration. These changes include the following.
```

Author: Fixed, thanks.

> __Improved ListBox Accessibility support__

Reviewer 3: Each of these should be a level-2 head (\#\#)

Author: Fixed, thanks.

> - Enabled UIA support for ListBox control.

Reviewer 3: Before the bulleted list, there should be a one-sentence introduction, like:

> On .NET Core, Windows Forms includes the following changes that affect the <xref:System.Windows.Forms.ListBox> control:

Author: Fixed, thanks.

> - Improved ListBox Accessibility support by adding the ScrollItem pattern to ListBox items and by enhancing the accessibility event raising and handling and Narrator navigation through the items (caps lock navigation is not correct and does not throw the navigation outside the control unintentionally).

Reviewer 3: ```suggestion
- Improved ListBox accessibility support by adding the ScrollItem pattern to ListBox items and by enhancing the accessibility event raising and handling and Narrator navigation through the items (caps lock navigation is not correct and does not throw the navigation outside the control unintentionally).
```
But the parenthetical expression seems like a non sequitur. Is it an exception? A qualification?

Author: Fixed, thanks.

__Improved ToolStrip Accessibility support__
- Improved navigation through ToolStrip items. Corrected ToolStrip and Menu shift-tab navigation (back-looping the menu items when shift-tab up-arrow is pressed - navigate to the bottom menu element)
- Improved Menu accessible navigation, corrected menu accessible control types for submenus to make submenus of type 'Menu' instead of 'MenuItem'.

__Improved PrintPreviewControl and PrintPreviewDialog Accessibility support__
- Improved accessible navigation (including Narrator navigation) through menu items.
- Improved High Contrast themes support and made the control element more contrasted.

__StringCollectionEditor Accessibility support__
- Added StringCollectionEditor with improved accessibility support.

__UI automation support for DataGridView, PropertyGrid, ListBox, ComboBox, ToolStrip and other controls__
Note: UI automation support is enabled for controls in runtime but is not used in design time.
For an overview of UI automation, see the [UI Automation Overview](https://docs.microsoft.com/dotnet/framework/ui-automation/ui-automation-overview).

>  

Reviewer 3: Lines 65 through 70 can be deleted. Optionally, the remaining commented-out list can be deleted.

Author: Fixed, thanks.
